### PR TITLE
[shardformer] llama support DistCrossEntropy

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#5175 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

llama support DistCrossEntropy

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/llama.py [hunk 1]
```diff
@@ -267,11 +273,18 @@ def llama_for_causal_lm_forward(
...
                 # Flatten the tokens
                 loss_fct = CrossEntropyLoss()
-                shift_logits = shift_logits.view(-1, self.config.vocab_size)
                 shift_labels = shift_labels.view(-1)
                 # Enable model parallelism
                 shift_labels = shift_labels.to(shift_logits.device)
-                loss = loss_fct(shift_logits, shift_labels)
+                if shard_config.enable_tensor_parallelism:
+                    tp_world_size = dist.get_world_size(shard_config.tensor_parallel_process_group)
+                    new_vocab_size = self.config.vocab_size // tp_world_size
```
Comment (ver217) [inline colossalai/shardformer/modeling/llama.py:281]: No need to compute again. The size of last dim of `shift_logits` is new vocab size

File: colossalai/shardformer/modeling/llama.py [hunk 2]
```diff
@@ -476,3 +490,107 @@ def forward(
...
+            shift_labels = labels[..., 1:].contiguous()
+            # Flatten the tokens
+            loss_fct = CrossEntropyLoss()
+            shift_labels = shift_labels.view(-1)
+            # Enable model parallelism
+            shift_labels = shift_labels.to(shift_logits.device)
+            if shard_config.enable_tensor_parallelism:
+                tp_world_size = dist.get_world_size(shard_config.tensor_parallel_process_group)
+                new_vocab_size = self.config.vocab_size // tp_world_size
+                shift_logits = shift_logits.view(-1, new_vocab_size)
```
Comment (ver217) [inline colossalai/shardformer/modeling/llama.py:578]: same

File: colossalai/shardformer/layer/loss.py [hunk 3]
```diff
@@ -78,17 +78,21 @@ def forward(ctx, vocab_logits: torch.Tensor, target: torch.Tensor, ignore_index:
         # calculate the loss
         # loss = log(sum(exp(x[i]))) - x[class]
         loss = torch.where(target == ignore_index, 0.0, torch.log(sum_exp_logits) - pred_logits)
-        loss = torch.sum(loss).div_(torch.sum(loss != 0.0))
+        non_zero_sum = torch.sum(loss != 0.0)
+        ctx.mean_grad = 1.0 / non_zero_sum
+        loss = torch.sum(loss).div_(non_zero_sum)
```
Comment (ver217) [inline colossalai/shardformer/layer/loss.py:83]: What does this mean? loss = sum * num_no_zero?

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/layer/loss.py (hunk 3 above)
Comment (flybird11111) [reply colossalai/shardformer/layer/loss.py:83]: I calculate the loss by loss = sum / num_no_zero.

File: colossalai/shardformer/layer/loss.py [hunk 4]
```diff
@@ -78,17 +78,21 @@ def forward(ctx, vocab_logits: torch.Tensor, target: torch.Tensor, ignore_index:
         # calculate the loss
         # loss = log(sum(exp(x[i]))) - x[class]
         loss = torch.where(target == ignore_index, 0.0, torch.log(sum_exp_logits) - pred_logits)
-        loss = torch.sum(loss).div_(torch.sum(loss != 0.0))
+        num_no_zero = torch.sum(loss != 0.0)
+        ctx.mean_grad = 1.0 / num_no_zero
```
Comment (ver217) [inline colossalai/shardformer/layer/loss.py:82]: `mean_grad`'s naming is weird

**Review by ver217 — APPROVED** [review]
